CUPS migration actors

zdohnal opened the pull request:
Hi!

I created two actors for cups migration - cupscheck and cupsmigrate. The code is currently 'working prototype', tested only manually for now.
I will move all 'non-actor' methods to outside module and create unit tests after my PTO, would you mind reviewing the code in meanwhile?

Thank you in advance!

Zdenek

artmello commented on repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py:
```diff
@@ -0,0 +1,17 @@
+from leapp.models import Model, fields
+from leapp.topics import SystemInfoTopic
+
+class Matrix(Model):
```
Could we have a more descriptive name for this model? Something like CupsMigrationMatrix? That way we can avoid some naming conflicts in the future. Also, some docstring documentation about the model can be helpful.

zdohnal replied:
No longer used.

artmello commented on repos/system_upgrade/el7toel8/models/cupsmigrationmodel.py:
```diff
@@ -0,0 +1,17 @@
+    env = fields.Boolean()
+    printcap = fields.Boolean()
+
+class Cupsmigrationmodel(Model):
```
ditto about having docstring documentation.

zdohnal replied:
Docstring added.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+
+Some changes happened only in one configuration file, so the last member of list is 'None'.
+"""
+cupsd_changes_list = [
```
cupsd_changes_list could be a Python's dictionary (or even namedtuples) what would simplify this code a bit. Instead of:
```python
INCLUDE = 0
(...)
cupsd_changes_list [['include', (...)]]
(...)
cupsd_changes_list[INCLUDE]
```
we can have:

```python
cupsd_changeslist {'include': ['include', (...)]}
cupsd_changes['include']
```
Also, since this is used at a single place afaict, moving it closer to where it is used (inside function) can help readability in my opinion.

zdohnal replied:
I chose different approach in the end.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+                      ['printcap', ['PrintcapFormat'], None, CUPSD_CONF, CUPSFILES_CONF]
+                    ]
+
+cupsfiles_changes_list = [
```
ditto for using a dictionary or namedtuples.

zdohnal replied:
Using different approach now.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+          with open(file, 'r') as conf:
+            content = conf.read()
+        except IOError:
+          return result
```
Would be interesting to return the error attached to the IOError exception. This can help us or users to understand better why an error happened.

zdohnal replied:
Done.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+            conf.write(new_content)
+          result = True
+        except IOError:
+          return result
```
ditto for returning a descriptive error.

zdohnal replied:
Corrected.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+
+      return result
+
+    def migrate_includes(self, include_files):
```
Same thing for this function, instead of returning boolean maybe we could return a descriptive error if something happened or empty string if everything is OK. Also, maybe would be a good practice to check if all files exists and are readable/writeable before doing proper changes. If the process fails in the end, but some changes were applied we can end up with a not working system.

zdohnal replied:
Now I check cupsd.conf and cups-files.conf for availability before doing anything else, exiting with error message if CUPS is installed but confs are missing.
Ad included files - their absence does not crash CUPS, just adds error into log - IMO it is user error to put non-existing file into configuration. It generates error (but service is still running) even on RHEL 7, so I chose to let 'Include' directives with non-existing files in cupsd.conf.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+      except IOError:
+        return False
+
+    def move(self, directives, source, target):
```
ditto about returning meaningful errors and doing all checks before properly executing something.

zdohnal replied:
Done.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+
+      return True
+
+    def process(self):
```
In this function there are some uses of `if variable is True:` and `if variable is False:`. We could use `if variable:` and `if not variable:` instead.

zdohnal replied:
Noted. Fixed.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+    def process(self):
+      for model in self.consume(Cupsmigrationmodel):
+        if model.migrateable is False:
+          report_generic(
```
If no change was made and the user should not take any action, I would advocate that we should not add an entry to the report. In that case I would remove this use of report_generic.

zdohnal replied:
No longer creates reports in this actor.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+                          summary='CUPS is not installed or its configuration is up-to-date.',
+                          severity='low'
+                        )
+          raise StopActorExecution('CUPS will not be migrated')
```
No need to raise an exception here, we could just use `return`.

zdohnal replied:
Ok.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+        if include is True:
+          result = self.migrate_includes(include_files)
+          if result is not True:
+            raise StopActorExecutionError('Include directive could not be migrated due IOError')
```
My other comments about functions returning meaningful error messages is related with this raised exception (and following ones). This error will stop Leapp processing (not only this actor), if that is what we want we should make it clear why it failed and how to solve the problem.

zdohnal replied:
Removed.

artmello commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/actor.py:
```diff
@@ -0,0 +1,339 @@
+          if result is not True:
+            raise StopActorExecutionError('PrintcapFormat directive could not be migrated due IOError')
+
+        report_generic(
```
If changes on CUPS are now transparent to user and the system will be working exactly it was before, I would remove this report_generic. If the changes made can have some impact into daily use of system I would add some detailed message about what was changed.

zdohnal replied:
No longer creates report in this actor.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+
+Some changes happened only in one configuration file, so the last member of list is 'None'.
+"""
+cupsd_changes_list = [
```
As mentioned on other file, I have the feeling that this can be better represented using a dictionary or namedtuples.

zdohnal replied:
Used different approach, hopefully more understandable.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+          if len(matched_list) > 0:
+            found = True
+            break
+        opened_config.close()
```
It is not necessary to close() opened_config. After the code exits current `with` context, the opened file will be closed automatically.

zdohnal replied:
Noted. Fixed.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+            for new_path in matched_list:
+              if new_path not in paths:
+                paths.append(new_path)
+          opened_config.close()
```
ditto for no need to close() opened_config.

zdohnal replied:
Noted, fixed.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+
+      return paths
+
+    def process(self):
```
No need for `is True` or `is False` at if statements.

zdohnal replied:
Corrected.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+            installed_cups = True
+
+      if installed_cups is False:
+        report_generic(
```
Since no action is needed, I would remove this report entry.

zdohnal replied:
Removed.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+                        )
+          migrateable = True
+        else:
+          report_generic(
```
Since no action is needed, I would remove this report entry.

zdohnal replied:
Removed.

artmello commented on repos/system_upgrade/el7toel8/actors/cupscheck/actor.py:
```diff
@@ -0,0 +1,226 @@
+        migration_matrix = self.look_for_affected_features(include_files_list=include_files_list)
+
+        if True in migration_matrix.values():
+          report_generic(
```
The process will happen (and succeed or fail) without user intervention. I think we can remove this entry from report and only notify user that migration was made.

zdohnal replied:
Removed.

zdohnal commented:
Hi everyone,

I'm still working on the actor for now, I'll let you know here in the ticket when it will be ready for new review.

bocekm commented:
Hi @zdohnal. Would you please try finishing this PR. Thanks.

zdohnal commented:
Ready for review after refactoring. Thank you for your patience!

pirat89 commented:
Thanks :)

pirat89 commented:
e2e tests

pirat89 commented:
@zdohnal Let's sync on monday. I think it is ok, but regarding UX we will need some changes in reports here. I can do all changes for you if you are run out of time regarding the point we missed that in previous reviews, the code seems clear to me.

zdohnal commented:
@pirat89 Ok, lets sync today or this week.

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+            reporting.Tags([reporting.Tags.DRIVERS]),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Remediation(hint=hint)
+        ]
```
note for me: keep separate

zdohnal replied:
@pirat89 my guess - the idea about 'groups' could be a aggregation of reports based on severity:

0) 'Drivers' report will stay separate
1) all other 'HIGH' reports will be merged in 1 func
2) all 'MEDIUM' reports will be merged together in 1 func

Let me know after you have time for looking into this.

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+                                      ' '.join(f for f in facts.include_files))
+        ]
+
+        report_func(args)
```
note for me: conf group 1. (severity none)

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+    :param obj facts: model object containing info about CUPS configuration
+    :param func report_func: creates report
+    """
+    title = ('PrintcapFormat directive is no longer in cupsd.conf')
```
note for me: conf group 2.

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+            reporting.RelatedResource('files',
+                                      ('/etc/cups/cupsd.conf '
+                                       '/etc/cups/cups-files.conf'))
+        ]
```
note for me: conf group 2.

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+            reporting.RelatedResource('file',
+                                      '/etc/cups/cups-files.conf')
+        ]
+
```
note for me: conf group 1. but:
- in case the cert and keys are in different directories, they will be copied into the `/etc/cups/ssl` directory (otherwise no action is needed). In such case, after change of keys or certs, both have to be updated, or what is recommended, to check your setup, modify scripts if needed to use the new directory and remove original ones maually. (msg draft)

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupscheck/libraries/library.py:
```diff
@@ -0,0 +1,204 @@
+            reporting.RelatedResource('file',
+                                      '/etc/cups/cupsd.conf')
+        ]
+
```
note for me: conf group 1.

AloisMahdal commented:
Note about tests: The `static` result is false positive (infra issue); the actual test passed and in a while (once I internally release morf-0.5.1) the part that caused problems will be disabled so you can then re-run tests.

OTOH, the `upgrade` part is valid; the upgrade went to emergency mode which led to timeout after total of 8h.  I spoke to @pirat89 about it; I'm assuming he's investigating it. (Please correct me if I'm wrong..)

pirat89 commented:
@AloisMahdal: You are right.

drehak commented:
leapp-ci build

AloisMahdal commented:
FTR, the failing tests above are probably (logs are garbage-collected already) for infra reasons that should be mostly fixed now.

bocekm commented:
@zdohnal, what is your plan with this PR? Is it WIP or do you need any addition review/help from the leapp team?

pirat89 commented:
@bocekm that's waiting for our investigation why the upgrade has been crashing.

zdohnal commented:
Hi @pirat89 ,

can I help anyhow at the moment?

AloisMahdal commented:
Can't say what to do with open peer review above, but there's been some progress on master in the meantime so rebasing & re-running `e2e tests` could be a good start...

AloisMahdal commented:
> @bocekm that's waiting for our investigation why the upgrade has been crashing.

@pirat89 not sure which crash you meant but if it was in beaker then logs are looong time gone.

zdohnal commented:
@AloisMahdal  @pirat89 Tests failed after rebasing to the current master with import errors like:

```
_ ERROR collecting repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/tests/test_digest_directive_check.py _

ImportError while importing test module '/payload/repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/tests/test_digest_directive_check.py'.

Hint: make sure your test modules/packages have valid Python names.

Traceback:

/payload/tut/lib/python2.7/site-packages/_pytest/python.py:507: in _importtestmodule

    mod = self.fspath.pyimport(ensuresyspath=importmode)

/payload/tut/lib/python2.7/site-packages/py/_path/local.py:704: in pyimport

    __import__(modname)

/payload/tut/lib/python2.7/site-packages/_pytest/assertion/rewrite.py:304: in load_module

    exec(co, mod.__dict__)

tests/test_digest_directive_check.py:3: in <module>

    from leapp.libraries.actor.library import digest_directive_check

E   ImportError: cannot import name digest_directive_check
```

Those errors didn't happen in previous testing. I checked for the error on the net and found a piece of information that it can be caused by __init__.py adding/removal, but no other actors I checked - quagga, bind, vsftpd - have it...
Would you mind helping me with it?

pirat89 commented:
@zdohnal Definitely. We will sync with you in the next two weeks. Just finishing works around the next release.

@AloisMahdal yes, it was in beaker.

zdohnal commented:
Hi @pirat89 ,

can we sync about the PR next week and figure out the plan what to do for merging it?

Thank you in advance!

Zdenek

pirat89 commented:
@zdohnal Hi, I am sorry I have forgotten about that completely. We had two very close releases recently. Yes, I will have definitely a time during the next week to sync with you. I have scheduled the mtg in our calendar.

pirat89 commented:
Discussed, we will go to push the work around Nov/Sep.

zdohnal commented on repos/system_upgrade/el7toel8/actors/cupsfeaturesfind/actor.py:
```diff
@@ -0,0 +1,29 @@
+from leapp.tags import FactsPhaseTag, IPUWorkflowTag
+
+
+class CupsFeaturesFind(Actor):
```
Renamed in the newest version.

zdohnal commented:
Ok, I updated the PR, fixed the actor based on @pirat89 findings (actor's libraries mustn't have a generic name, otherwise the import will fail) and incorporated two of suggested changes - renaming cupsfeaturesfind to cupsscanner and using StopActorExecution error if conf files are missing.

@pirat89 then the only thing remaining is a conf groups, which I don't recall exactly what supposed to mean - my guess is an aggregation of reports either by severity, tag or number of affected files, but I'm not sure. If you manage to check it earlier and write down the expectations, I can try to update the code accordingly.

Ad failed e2e test: @AloisMahdal would you mind helping me with e2e test [results](https://baseos-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/oamg-ci-test-internal_copr-ts_sti-artemis/213/artifact/consoleText.txt.gz)? I see there a host cannot be reached, is there something I can do?

mkluson commented:
e2e test fail: the machine did not boot after the upgrade, I will rerun it once again with different parameters to get serial console log

AloisMahdal commented:
Still failing.. Not sure why.  I'm running Beaker-based Murphy suite as backup or for comparison.

AloisMahdal commented:
Murphy upgrade test got stuck too, because there's traceback & emergency mode:

        [..]
        [..]
    [  784.597300] upgrade[593]: 2020-11-10 08:39:05.443 DEBUG    PID: 839 leapp.fallback: Updating Vim configuration file /etc/virc. 
    [  784.647614] upgrade[593]: 2020-11-10 08:39:05.495 INFO     PID: 3 leapp.workflow.Applications: Executing actor migrate_brltty 
    [  784.705840] upgrade[593]: 2020-11-10 08:39:05.553 INFO     PID: 3 leapp.workflow.Applications: Executing actor cupsfilters_migrate 
    [  784.801243] upgrade[593]: 2020-11-10 08:39:05.648 INFO     PID: 3 leapp.workflow.Applications: Executing actor cups_migrate 
    [  784.854298] upgrade[593]: Process Process-183: 
    [  784.855407] upgrade[593]: Traceback (most recent call last): 
    [  784.856554] upgrade[593]:   File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap 
    [  784.858127] upgrade[593]:     self.run() 
    [  784.859028] upgrade[593]:   File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run 
    [  784.860519] upgrade[593]:     self._target(*self._args, **self._kwargs) 
    [  784.861731] upgrade[593]:   File "/root/tmp_leapp_py3/leapp/repository/actor_definition.py", line 72, in _do_run 
    [  784.863337] upgrade[593]:     actor_instance.run(*args, **kwargs) 
    [  784.864450] upgrade[593]:   File "/root/tmp_leapp_py3/leapp/actors/__init__.py", line 335, in run 
    [  784.865996] upgrade[593]:     self.process(*args) 
    [  784.868378] upgrade[593]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/cupsmigrate/actor.py", line 21, in process 
    [  784.870180] upgrade[593]:     cupsmigrate.migrate_configuration() 
    [  784.871326] upgrade[593]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py", line 180, in migrate_configuration 
    [  784.873383] upgrade[593]:     facts = consume_function(CupsChangedFeatures) 
    [  784.874685] upgrade[593]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py", line 164, in _get_facts 
    [  784.876685] upgrade[593]:     return next(api.consume(model)) 
    [  784.877813] upgrade[593]: StopIteration 
    [  784.889279] upgrade[593]: ============================================================================================= 
    [  784.891192] upgrade[593]: Actor cups_migrate unexpectedly terminated with exit code: 1 - Please check the above details 
    [  784.892879] upgrade[593]: ============================================================================================= 
        [..]
        [..]


That was with:

 *  *leapp-0.11.1-100.202011021301Z.90b67be.master.el7_6*
 *  *leapp-repository-0.12.0-0.202011100708Z.950cddf.cups_migration.PR181.el7_6*

AloisMahdal commented:
Note that the above was clean installation, ie. cups was not installed.

zdohnal commented:
Aha, the default value was missing from ```next()``` call in ```cupsmigrate``` actor - ```None``` :). Fixed now and the testsuite was updated to check this case.

Rezney commented:
@zdohnal @pirat89 this PR seems to be "a little" stuck. Would you have time to get it back on track if it is still valid?

pirat89 commented:
I will look at it today/tomorrow.

pirat89 commented:
/rerun

pirat89 commented:
/rerun

pirat89 commented on repos/system_upgrade/el7toel8/actors/cupsmigrate/libraries/cupsmigrate.py:
```diff
@@ -0,0 +1,232 @@
+        except IOError as error:
+            error_list.append('PrintcapFormat directive: {}'.format(error))
+
+    if error_list:
```
@zdohnal we should ad here `TODO: add post upgrade report`. As we are not doing post-upgrade reports so much yet, we do not have to add it now. But we should add least add TODO here.

pirat89 commented:
@zdohnal so far it seems ok and working. However I am still confused about the latest change "startswith" -> "rsplit". Probably I am working too long today already. Let's sync tomottor. I've tested the solution for `PrintcapFormat` and everything has been done as expected.

And I want to discuss the TLS one more time in cupsd (I mean, `Digest` -> `Basic` change).

zdohnal commented:
> @zdohnal so far it seems ok and working. However I am still confused about the latest change "startswith" -> "rsplit".

To sum it up, "startswith" is 'vulnerable' to substrings - f.e.:

- server.key lives in /etc/cups/ssl
- server.cert lives in /etc/cups/ssl/smdir

'startswith', if the former gets into certkey_values first, then the latter matches as substring and the function incorrectly puts ServerKeychain /etc/cups/ssl and no copying happens -> cupsd won't be using the correct cert.

If we stayed with 'startswith', we would need to 'preprocessed' certkey_values or choose the value with the longest path instead of the first value in the list. In that case 'rsplit' function looked cleaner to get the filename and its path, and than do the exact comparison.

> And I want to discuss the TLS one more time in cupsd (I mean, `Digest` -> `Basic` change).

Ok, basically how the users do the remote auth doesn't change - the thing which changes is how 'protected' are creds - with Digest, the creds were hashed by hashing alg and sent over unencrypted connection. with Basic, the creds are in plain text, but the connection is protected by TLS, thus not that easy to leak.

pirat89 reviewed (approved):
It's here!!! Waiting for QE approval!

pirat89 commented:
/rerun

pirat89 reviewed (approved):
just regular rebase, everything is same as before. re-ack

pirat89 commented:
btw, in case the e2e fails, ignore it. the test is broken nowadays and it's going to be removed soon.

zdohnal commented:
\o/ thx for merging it :)

pirat89 commented:
Thanks for all your hard work you did in this PR!!
